source/alloc/src/heap.rs

```diff
@@ -1,505 +1,128 @@
-            self.heap_wait.wait().await.unwrap();
-        }
-    }
-
-    pub async fn allocate_arc<T>(&'static self, mut item: T) -> HeapArc<T> {
-        loop {
-            // Is the heap inhibited?
-            if !self.inhibit_alloc.load(Ordering::Acquire) {
-                // Can we get an exclusive heap handle?
-                if let Ok(mut hg) = self.lock() {
-                    // Can we allocate our item?
-                    match hg.alloc_arc(item) {
-                        Ok(hb) => {
-                            // Yes! Return our allocated item
-                            return hb;
-                        }
-                        Err(it) => {
-                            // Nope, the allocation failed.
-                            item = it;
-                        }
-                    }
-                }
-                // We weren't inhibited before, but something failed. Inhibit
-                // further allocations to prevent starving waiting allocations
-                self.inhibit_alloc.store(true, Ordering::Release);
+            None => {
+                debug_assert!(false, "Deallocating a null?");
+                return;
             }
-
-            // Didn't succeed, wait until we've done some de-allocations
-            self.heap_wait.wait().await.unwrap();
         }
     }
+}
 
-    pub async fn allocate_array_with<F, T>(&'static self, f: F, count: usize) -> HeapArray<T>
-    where
-        F: Fn() -> T,
-    {
-        loop {
-            // Is the heap inhibited?
-            if !self.inhibit_alloc.load(Ordering::Acquire) {
-                // Can we get an exclusive heap handle?
-                if let Ok(mut hg) = self.lock() {
-                    // Can we allocate our item?
-                    //
-                    // NOTE: We borrow `f`, since `&Fn() -> T` still impls `Fn() -> T`, and allows
-                    // us to potentially call it multiple times.
-                    match hg.alloc_box_array_with(&f, count) {
-                        Ok(hb) => {
-                            // Yes! Return our allocated item
-                            return hb;
-                        }
-                        Err(_) => {
-                            // Nope, the allocation failed.
-                        }
-                    }
-                }
-                // We weren't inhibited before, but something failed. Inhibit
-                // further allocations to prevent starving waiting allocations
-                self.inhibit_alloc.store(true, Ordering::Release);
-            }
+#[cfg(feature = "use-std")]
+impl UlAlloc for std::alloc::System {
+    const INIT: Self = std::alloc::System;
 
-            // Didn't succeed, wait until we've done some de-allocations
-            self.heap_wait.wait().await.unwrap();
-        }
+    unsafe fn init(&self, _start: NonNull<u8>, _len: usize) {
+        panic!("Don't initialize the system allocator.");
```

Comment:
would it make more sense for this to just be a nop rather than panicking, so we can have cross-platform kernel init code use it? if we're not going to call `init` in a generic context, we could also just make it be a method on the particular allocator types, rather than a trait method...